developer_docs/type_inference.md

```diff
@@ -43,3 +43,35 @@ These objects contain all the information necessary to create a Variable from a
+
+#### Fixed Size Type
+A `FixedSizeType` is an object whose size in memory is known and cannot change from one instance to another (e.g. `float64`, `int32`, `void`). In most cases the developer will need the sub-class `FixedSizeNumericType` which refers to the subset of fixed size types which contain numeric values. These objects are characterised by a `primitive_type` describing the category of datatype (integer/floating point/etc) and a `precision`. They additionally implement two magic methods:
+-   `__add__` (+)
+-   `__and__` (&)
+
+The add operator describes what happens when two numeric types are combined in an arithmetic operator. In almost all cases this is sufficient to describe all resulting datatypes. Special cases (e.g. float for integer division) are handled in the associated operator in `ast.operators` or `ast.bitwise_operators`.
+
+The and operator describes what happens when two numeric types are combined in a bitwise comparison operator. This only applies to integers and booleans.
+
+When using these operators on an unknown number of arguments it can be useful to use `NativeGeneric()` as a starting point for the sum.
+
+#### Container Type
+A `ContainerType` is an object which is comprised of `FixedSizeType` objects (e.g. `ndarray`,`list`,`tuple`, custom class). The sub-class `HomogeneousContainerType` describes containers which contain homogeneous data. These objects are characterised by an `element_type`. The elements of a `HomogeneousContainerType` are instances of `PyccelType`, but they can be either `FixedSizeType`s or `ContainerType`s.
+
+`HomogeneousContainerType`s also contain some utility functions. They implement `primitive_type` and `precision` to get the properties of the internal `FixedSizeType` (even if that type is inside another `HomogeneousContainerType`). They also implement `switch_basic_type` which creates a new `HomogeneousContainerType` which is similar to the current `HomogeneousContainerType`. The only difference is that the `FixedSizeType` is exchanged. This is useful when we want to preserve information about the container but need to change the type. For example, when we divide an integer by another we get a floating point type. When we divide a NumPy array or a CuPy array of integers by an integer (or array of integers) we get a NumPy/CuPy array of floating point numbers (with default Python precision). In order to preserve the container type we therefore call `switch_basic_type`.
```

Comment:
The last sentence in this paragraph is not totally clear to me. What do you mean with "preserve the container type"?

Reply:
I mean that if I have a `list[np.ndarray[float]]` and I call `switch_basic_type` then I still retain the information about the container type, ie. `list[np.ndarray`. If I try to rebuild the type manually then I will have to avoid the trap of getting `list[float]`. After the next big PR this will also be about preserving information saved in the container type. `np.ndarray` will contain information about the rank/order so building a new `np.ndarray` with a different type would involve extracting this information to pass it into the new type. Obviously this would be a pain to maintain if we add new parametrisations.

Reply:
In this context I mean in order to get `np.ndarray[int]` instead of just `int`

Reply:
Thanks for the explanation! Could you add an example to the docs? I feel that would be very helpful.

Reply:
There is already an example in the second to las sentence. I have added an extra sentence to try and help : 7314db11d